File: src/main/scala/booklet/pure/Failure.scala
package booklet.pure

case class Failure(message: String, cause: Option[Throwable])

object Failure {
  def fromThrowable(t: Throwable): Failure = Failure(t.getMessage, Some(t))

  def fromDecodingException(s: String): Failure = Failure(s"Json decoding exception: $s", None)
}

File: src/main/ts/booklet/pure/Failure.ts
export class Failure {
  message: string;
  cause: Error | null;

  constructor(message: string, cause: Error | null = null) {
    this.message = message;
    this.cause = cause;
  }

  static fromThrowable(t: Error): Failure {
    return new Failure(t.message, t);
  }

  static fromDecodingException(s: string): Failure {
    return new Failure(`Json decoding exception: ${s}`);
  }
}
